changes/2024-TBD_cache-across-hierarchy-keyrings/change.md

```diff
@@ -0,0 +1,454 @@
+can only ever call KMS with one KMS Relationship,
+which is, at least partly,
+configured on the KMS Client determined
+at the Hierarchy Keyring's construction.
+
+By KMS Relationship, we mean any or all of the following:
+
+- KMS Configuration
+- Credentials used when creating the KMS Client, and thus
+  used when calling KMS
+- Other properties of the KMS Client, such as the region,
+  or request headers
+
+The Local Cryptographic Material Cache of
+the Hierarchy Keyring instance is then only
+populated with Branch Keys that correspond with
+that KMS relationship.
+
+Which is appropriate,
+as it is clear under what KMS relationship
+a Branch Key is accessed.
+
+However,
+the Hierarchy Keyring,
+and it's Keystore,
+have a runtime cost,
+exerting memory pressure
+and, without manual optimization,
+requiring at least 2 TLS handshakes
+when first serving a request
+(TLS to KMS & TLS to DDB).
+
+Additionally,
+the local Cryptographic Materials Cache
+exerts some runtime cost,
+particularly in a multi-threaded environment,
+when a background worker thread MAY be refreshing
+or pruning entries of the cache.
+
+For MPL Consumers that MUST work with Branch Keys
+under different KMS Relationships,
+this runtime cost adds up.
+
+These MPL Consumers MAY end up establishing
+a LRU Cache of Hierarchy Keyrings.
+Which, while workable, is sub-optimal,
+and clearly makes the Hierarchy Keyring,
+in these conditions,
+"Hard to Use".
+
+The objective, with these changes,
+is to make the Hierarchy Keyring
+"Easy to Use" in a multiple KMS Relationship
+environment.
+
+## Summary
+
+Relevant [sim](https://sim.amazon.com/issues/CrypTool-5311).
+
+To improve the MPL Consumer data key caching experience,
+when using the Hierarchy Keyring,
+we need to allow caching across KeyStores/KMS Clients/KMS Keys.
+
+To facilitate Caching across KeyStores/KMS Clients/KMS Keys,
+we MUST break the Cryptographic Materials Cache (CMC)
+out of the Hierarchy Keyring.
+
+By allowing MPL Consumers to provide an already initialized CMC
+to the Hierarchy Keyring at construction,
+the CMC MAY cache Branch Keys protected by different
+KMS Relationships.
+
+This simplifies Multiple KMS Relationship MPL Consumers,
+as they do not need to stand up a LRU Cache of Hierarchy Keyrings.
+
+Instead, they may maintain one CMC.
+They still create a Hierarchy Keyring instance per KMS Relationship,
+and they MUST use the correct Keyring to retrieve material
+from the Cache.
+
+But they only need to maintain the common cache.
+
+In the future, the CachingCMM will be introduced to
+Crypto Tool's Dafny products.
+The CachingCMM and the Hierarchy Keyring both consume CMCs.
+Thus, it will be possible to provide a CMC to both a
+Hierarchy Keyring & a CachingCMM.
+This cache identifier design MUST consider
+how the cache entries of multiple Hierarchy Keyrings
+and CachingCMMs will be appropriately separated.
+
+## Requirements
+
+- We must allow caching across KeyStores/KMS Clients/KMS Keys
```

Comment:
nit: We have switched from KeyStore/keystore to Key Store. 
Two words.

Reply:
Okay, thanks!